modules/nf-core/bbmap/filterbyname/main.nf

```diff
@@ -0,0 +1,53 @@
+process BBMAP_FILTERBYNAME {
+    tag "$meta.id"
+    label 'process_single'
+
+    conda (params.enable_conda ? "bioconda::bbmap=39.00" : null)
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/bbmap:39.00--h5c4e2a8_0':
+        'quay.io/biocontainers/bbmap:39.00--h5c4e2a8_0' }"
+
+    input:
+    tuple val(meta), path(reads)
+    val output_extension
```

Comment:
This could be set in the modules.config, like [samtools view ](https://github.com/nf-core/modules/blob/master/modules/nf-core/samtools/view/main.nf) does.

Reply:
I do not think filterbyname.sh offers an option to pass a file extension as a flag (samtools view has `--output-fmt`). I could do all manipulations with it and then remove it from `args`, but I think it just overcomplicates things. What do you think?

Here is what I would have to do:
```
// if `output-fmt` was passed -> use it, otherwise treat everything after the
// first comma as prefered `output-fmt`
def file_type = args.toLowerCase().contains("--output-fmt fasta") ? "fasta" :
                    args.toLowerCase().contains("--output-fmt fa") ? "fa" :
                    args.toLowerCase().contains("--output-fmt fastq") ? "fastq" :
                    args.toLowerCase().contains("--output-fmt fq") ? "fq" :
                    meta.single_end ? reads.name.tokenize(".")[1..-1].join(".") :
                    reads[0].name.tokenize(".")[1..-1].join(".")

args = args.replaceAll("--output-fmt \\w+ ?", "")
```

Reply:
Does the tool define the output format by this extension on the output file? Or is the output the same as the input?

Reply:
Then I think using a dictionary `args` would be the best choice, see here for examples:
[nextflow.config](https://github.com/nf-core/modules/blob/2edb02c68b0d2e3b31dca6360f4a683d5b0f2970/tests/modules/nf-core/ascat/nextflow.config#L6-L15)
[main.nf](https://github.com/nf-core/modules/blob/2edb02c68b0d2e3b31dca6360f4a683d5b0f2970/modules/nf-core/ascat/main.nf#L34-L50)

Also don't forget to specify that the user should use it like this inside of the meta.yml, like [this](https://github.com/nf-core/modules/blob/2edb02c68b0d2e3b31dca6360f4a683d5b0f2970/modules/nf-core/ascat/meta.yml#L17-L40)

Reply:
How would it work with passing `args` to the command itself (line 43)? Do you want me to manually write out every single CLI option available to solve that?

Reply:
No, just the extension. You could add `args` again inside of `args` and use that for the CLI arguments :). (I should have been more clear on that sorry)

Reply:
Okay, I am almost there. Here is what I have now:
```
def args               = task.ext.args   ?: [ args:"" ]
def prefix             = task.ext.prefix ?: "${meta.id}"
def output_extension   = args.output_extension ? "$args.output_extension" :
    meta.single_end ? reads.name.tokenize(".")[1..-1].join(".") :
    reads[0].name.tokenize(".")[1..-1].join(".")
def interleaved_output = true ? "$args.interleaved_output" : false

def input            = meta.single_end ? "in=${reads}" : "in=${reads[0]} in2=${reads[1]}"
def filtered         = meta.single_end || interleaved_output ? "out=${prefix}.${output_extension}" :
    "out1=${prefix}_1.${output_extension} out2=${prefix}_2.${output_extension}"

```

What should I put for output `tuple val(meta), path("*.${output_extension}"), emit: reads`? Because user can specify any extension and leaving it as is does not work (output_extension gets defined later)

Reply:
That should work if you remove the `def` in front of `output_extension` (not 100% sure on this, but I remember it like this)

Reply:
Removing `def` gives me a `Module compilation error` :(

I have tried defining `output_extension` outside the `script` block and changing `args.output_extension` to `ext.args.output_extension`, but from what I understand, config is not read in yet at that point, so I am getting `null` as a value. I was also toying with an idea of grabbing anything that is not `versions.yml` or `*.log` with regex, but group negation within a glob is not a thing from what I have learned (only character negation through `[!<SOME-CHARACTER(s)>]` is available).

I'll poke around more, but I welcome any ideas, unless we want to just stick with `output_extension` being one of the inputs :)

Reply:
Like they do [here](https://github.com/nf-core/modules/blob/202683bfc98ddecdd456ea73268e330bca2e5c5a/modules/nf-core/samtools/convert/main.nf#L16)

Reply:
I have tried that, but I do not think typing out every single possible extension written with different capitalizations is the best solution (capitalization is important because if `output_extension` is not passed, it uses input's extension over which I have no control).

Reply:
I have two branches reads:
[One that uses `args` ](https://github.com/tokarevvasily/nf-core-modules/tree/bbmap-filterbyname-config). It has a bunch of predefined outputs, but I foresee issues if user does not specify output extension that they want and input's extension does not match any of the ones that I pass.

Here is [another branch that uses `input val`](https://github.com/tokarevvasily/nf-core-modules/tree/bbmap-filterbyname-vals). I like this approach significantly better and there is already a similar use case in [fastp](https://github.com/nf-core/modules/blob/202683bfc98ddecdd456ea73268e330bca2e5c5a/modules/nf-core/fastp/main.nf#L12), where `val save_trimmed_fail` is not directly used as an input for the tool, but rather controls how it runs.

For both I have added an option for interleaved output when using paired-end data!
Let me know which one you want me to push here!

Reply:
I found a solution for the earlier problem with using `output_extension` in the output definition => see how they used `prefix` [here](https://github.com/nf-core/modules/blob/master/modules/nf-core/krakenuniq/build/main.nf)



> [One that uses `args` ](https://github.com/tokarevvasily/nf-core-modules/tree/bbmap-filterbyname-config). It has a bunch of predefined outputs, but I foresee issues if user does not specify output extension that they want and input's extension does not match any of the ones that I pass.

This is the one that follows the guidelines most closely. So I would suggest this one.



> Here is [another branch that uses `input val`](https://github.com/tokarevvasily/nf-core-modules/tree/bbmap-filterbyname-vals). I like this approach significantly better and there is already a similar use case in [fastp](https://github.com/nf-core/modules/blob/202683bfc98ddecdd456ea73268e330bca2e5c5a/modules/nf-core/fastp/main.nf#L12), where `val save_trimmed_fail` is not directly used as an input for the tool, but rather controls how it runs.

For this, an exception can be made, but that's not something I can decide by myself. Can you ping `@modules-team` in Slack and ask for some more opinions on this?

Reply:
(I'll post a link to this in Slack)

Reply:
Is there a reason not to just stick with `fasta.gz` as the file suffix? Output should always be compressed.

Reply:
@nvnieuwk I have tried this before and that's where I was getting `Module compilation error`. However, since your example clearly worked, I did some additional digging and realized that once I define a variable without `def` nothing after it can have `def`. I made appropriate changes, and everything worked! I am not sure if this behavior is expected, but it is definitely new to me.

As a note, I have also updated BBmap to the most recent version, however, singularity container is not yet on `https://depot.galaxyproject.org/`, but the new version is on `conda` and `quay.io/biocontainers`. As a result, tests with singularity will fail until new image gets uploaded.

Reply:
Awesome, I didn't know that myself :)

The singularity container should be there soon (it usually takes a bit longer than the docker containers)